### ISSUE TITLE
Implement an inspector "Action Area"

### ISSUE THREAD

**Issue by reduz** [godotengine/godot-proposals#4273]
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

This has been a running problem with Godot UX since its inception. Due to the inspector originally being a tree it was not really possible to put anything in there (as other engines like Unity often does) so many add-ons and tools had to be placed somewhere else. 

![image](https://user-images.githubusercontent.com/6265307/159881444-e07a08b7-2678-43bc-953c-e76246554259.png)

Beginning Godot 3.1, the inspector now can use plugins and can generally be more customizable. The problem is that, due to how things evolved, the inspector has become strictly a tool to _edit data_. As such, anything related to performing actions related to the nodes being edited was moved to the top bar, such as this, be it 2D, 3D or other nodes:

![image](https://user-images.githubusercontent.com/6265307/159881809-b05d7ac7-79c3-4966-b37b-13908790e6bd.png)
![image](https://user-images.githubusercontent.com/6265307/159883308-c4f2d6f6-b152-4b72-81e0-197810d7e1be.png)
![image](https://user-images.githubusercontent.com/6265307/159883698-2697a39b-e07a-492b-974d-228320f43b86.png)
![image](https://user-images.githubusercontent.com/6265307/159883765-51464e09-c8a8-4bc8-a690-1abf3742e8a7.png)

To understand better what the problem is, look at this screenshot. The disconnect between the object properties and the action makes it difficult to understand that both are related. Many users feel lost because they simply don't notice the actionable objects appearing in the toolbar when selecting a node:

![image](https://user-images.githubusercontent.com/6265307/159886264-eb2b13b3-dd68-454c-83c2-b8bb5b9da28f.png)

The list goes on, there are dozens of these editors for 2D and 3D. One alternative for this is using Drawers (#55630), but they have a couple of issues:
* They mix action and editing, hence its inconsistent.
* They can't really be used in 3D because the 3D viewports can be split and moved around so drawers can be opaqued.
* Its still a bit separate
* Folded drawers are not entirely clear
* Generally the Godot UI tries to keep away from floating objects for simplicity.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The solution to me would be to implement "action areas". This means having an area with an icon of the type of object and the actionable objects.  They would require to be a single line to not take much space, but if necessary they can expand and collapse internally (not universally).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Examples of action areas (note that this includes simplifying the inspector header a bit):

![image](https://user-images.githubusercontent.com/6265307/159890359-333dbd2d-a619-489a-a3a9-6d79d2bf8e50.png)
![image](https://user-images.githubusercontent.com/6265307/159891313-bdfcc104-9f0c-407d-a7e9-03246a8bbc82.png)
![image](https://user-images.githubusercontent.com/6265307/159897829-4e753504-58c7-43ea-974f-6ab9a9a90177.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

**Comment by SirLich** [comment]
I strongly agree with your "problem statement". You've correctly identified that the action-bar is too subtle, and often missed. I say that as a relative newcomer to the engine, after experience with others. 

I feel less confident in your "solution statement", however. 

My fear is that the inspector is already being used for too many things. I personally find it very overwhelming and cramped feeling, compared to, say, the UE4 Properties Panel.

Conceptually, I think a more reasonable thing would be to introduce another panel:
 - Inspector (properties)
 - Node (don't like this name, but fine)
 - MultiMesh (name header w/ the *class* of the selected node). 

Benefits:
 - Reduce ambiguity for things like "filter properties"
 - Allow these actions to take more space! Instead of just icons, you can put a whole button with text there!

Cons:
 - Requires flipping back and forth between panels?

**Comment by Jojox** [comment]
I like this Action Area but I'm not sure if the deployed menu and more complexe behavior (adding point to a line for example) will have enough space to be usable. Could we use a similar tab than the one for connecting signal ? I feel like this would feel natural to use now I got use to the import and signal tabs. Codewise this would isolate this action panel and allow more complex plugin.

**Comment by reduz** [comment]
@SirLich In my view, If something is getting too crowded, the solution is not to add even more sections. Instead, we should simply make it _less_ crowded. This is why, if you look in detail in this proposal, it already includes some simplifications to the inspector to ensure the amount of elements and size remains the same.

**Comment by SirLich** [comment]
> In other words, the primary priority is that things make sense, avoiding the crowding is secondary.

For me, it "makes sense" to separate actions from configurable properties, but maybe I'm misunderstanding what options are currently available in the contextual "toolbar menu" (the one you want to move into the inspector. 

The nice thing about placing things into windows, is you leave the door open for shuffling these around, or floating them.

**Comment by Jojox** [comment]
For me, the feature breaker is:
![image](https://user-images.githubusercontent.com/6182189/159909076-34e131c3-6750-407b-a605-dfc405d13c83.png)
The fact that this menu can create new things in the scene tree makes me thing that it does not fit in the "inspector" panel.

**Comment by TackerTacker** [comment]
Hmmm... I get that a list with a bunch of slots for different items is much more appealing from an engineering point of view, but I have to say, just simply from an UX point of view the contextual drawers proposal looks a lot more intuitive, cleaner, less cluttered and more appealing imo. 
Both are better than the current solution however.

**Comment by gaudecker** [comment]
Personally I like that the inspector is mostly for editing data. I think https://github.com/godotengine/godot/pull/55630 provides the facilities where these actions should live. We just need a proper plugin API to add these categories and actions to the drawer.

Blender also has these drawer type UI elements, which (despite their discoverability in Blender) are quite nice to use.
![Screenshot from 2022-03-24 13-55-53](https://user-images.githubusercontent.com/872609/159911628-e04674dc-6d4b-435d-8129-e967b76c35c3.png)

**Comment by reduz** [comment]
@gaudecker Right, but you are contradicting yourself. Both the drawers in Blender that you make reference to and the functionality in https://github.com/godotengine/godot/pull/55630 are _also_ for editing data, which is much better covered IMO in #4274 for the Godot use case.

Actions in Godot are not analogous to the drawers in Blender, they are for making things happen, not editing values. They are also very simple actions contextual to the object being edited, which is why having a small action area makes more sense in this case.

**Comment by reduz** [comment]
@TackerTacker This one is not as much as a replacement to contextual drawers as #4274 is. Contextual drawers are for the most part for editing data. This is for actions, which are not replaced by them in that PR.

**Comment by gaudecker** [comment]
They are, but they don't have to be, and actions sometimes end up editing data by side effect, so they are similar in that respect

**Comment by reduz** [comment]
@gaudecker In this case, for the most part, they don't really edit data.

**Comment by Jojox** [comment]
@reduz I agree with the visibility issue, the point by @SirLich about the ability to organize the floating window is good too.
Thinking of something like this by default to solve the visibily issue:
![image](https://user-images.githubusercontent.com/6182189/159914905-909d1e2f-0b42-4dff-bb2b-2f42c6817ad6.png)
Then it's up to the user to arrange it for his needs, i feel like this has the same feeling than the action area but the new panel give the advantage for more advanced usage and flexibility for user.

**Comment by reduz** [comment]
@gaudecker 

> I see the drawers as containers for tools. 

As I mentioned here and in other places, the whole concept of drawers does not work in 3D because you can split the viewports there and this will cover other buttons or interfaces. Additionally, they are not consistent with the rest of the Godot interface.

**Comment by reduz** [comment]
@Jojox If Godot used a Blender or Unity like UI, your proposal would make sense, but it does not. The general vision in Godot is that things come into view when you need them. The vision in Blender or Unity is that you decide what goes into view.

**Comment by TackerTacker** [comment]
> @TackerTacker This one is not as much as a replacement to contextual drawers as #4274 is. Contextual drawers are for the most part for editing data. This is for actions, which are not replaced by them in that PR.

I'm confused then, this is the first examples you show 
![](https://user-images.githubusercontent.com/6265307/159881809-b05d7ac7-79c3-4966-b37b-13908790e6bd.png)
and this is the first example the contextual drawers proposal shows
![](https://user-images.githubusercontent.com/11782833/144729290-6b3524b4-6ed4-4879-a0d4-1cb6664a6d3b.png)
are these two proposals not trying to solve the same issue?

**Comment by KoBeWi** [comment]
The toolbar was tweaked to save space. We just went with a quicker solution for now, because it was important (there were 6 issues about it).

**Comment by Zireael07** [comment]
@reduz What I like in @Jojox's idea since it places all actions in a visibly labeled place/container unlike the OP mockup which just shoehorns them next to the search box (which might confuse people into thinking the search box is for actions). 

Maybe put the Actions container above the search box to make it clear?

**Comment by reduz** [comment]
@Zireael07 Yeah I guess no harm placing it over the search box

**Comment by SilencedPerson** [comment]
> @reduz What I like in @Jojox's idea since it places all actions in a visibly labeled place/container unlike the OP mockup which just shoehorns them next to the search box (which might confuse people into thinking the search box is for actions).
> 
> Maybe put the Actions container above the search box to make it clear?

Text is read from top to bottom, left to right. UI should follow that. If something is under the search bar, it should be assumed it is a separate thing. Though i agree there should be more explicit separation. Nevertheless, search bar should always be the top most thing in piece of UI anywhere.

**Comment by tinmanjuggernaut** [comment]
For years I never noticed those extra action buttons up top, even writing custom functionality to duplicate what already existed (MMI placement). I've always expected actions in the inspector and in my own tool scripts I setup boolean actions that upon click, perform actions using the set/get functions. I would still like to see actions at the top of the inspector.

Or a separate actions panel, and a more flexible window layout so I could put actions above the inspector.

Also the top bar gets crowded by plugins. Pushing all that to another panel or inspector section will clean all that up.

IMO, The top bar should instead be limited to viewport functions with things like top/left/right, info, fps, render mode should be exposed to it.

**Comment by KoBeWi** [comment]
btw things like drawers can likely be implemented in GDScript as a custom plugin (if not, we could make sure it's possible), so you can customize the editor to have dedicated editors for everything. If something is not in core, it doesn't mean you can't have it.

**Comment by Ansraer** [comment]
Honestly I am not certain if I want those buttons in the Inspector area. That place already is far too crowded as is.
It also feels somewhat unintuitive to me. In my mind the inspector is there to adjust variables, not to click buttons.

Ideally we would place it in its own area and use a system similar to ImGuis docking to allow users to adjust the editors UI layout to their liking. If they want they can move it into the inspector, make it their own tab or even move it to another screen. But I guess thats out of scope for now.

What I really liked about the drawer proposal is how small the tools are when they are collapsed and not needed, while still being close to where you need them. Plus, it feels similar to how blender does things, which is nice for anyone using both.

I think we should start by looking at every menu item that is currently show in that menu bar and ask ourselfes how often users need to use it. I, for example, don't need to bake VoxelGI every few minutes, so that button could be moved to a sub menu somewhere else (or just the command search, if that has already been merged).

**Comment by reduz** [comment]
@Ansraer
> That place already is far too crowded as is.
It also feels somewhat unintuitive to me. In my mind the inspector is there to adjust variables, not to click buttons.

IMO as I mentioned before thats entirely subjective thinking. This is a practical solution because if you move these somewhere else, discoverability is terrible. Even in drawers (which would be next to the inspector anyway :shrug: ) discoverability would be worse than with this proposal.

What is important when creating software is finding what works best with the current resources available, not figuring out the best conceptual subjective designs that feel good in your head.

**Comment by farfalk** [comment]
@reduz completely agree with the discoverability part. Not joking, when I started (before the useful highlighting that was recently introduced) I completely ignored the fact that some nodes had additional options in the toolbar. Couldn't design UI's because of it, heh!

Yet... the fact that it's called "Inspector" makes me feel that it evolved as a thing to "inspect" more than "act", and is (mostly) structured as it is, with modular and cascading properties, because of its evolved and agreed upon role.

Ironically, to me it semantically makes more sense to add it to the "Node" panel, because "you can take those actions with that node". But that would probably be even worse. The Node panel is already controversial, and completely ignored by those who don't use groups or signals from the editor but only from code.

**Comment by mrjustaguy** [comment]
How about allowing users to tag uncommonly used buttons in the inspector and putting them in another inspector tab? By that I Mean here with Node
![image](https://user-images.githubusercontent.com/50302479/159949680-a29819b1-6257-4d32-9d67-a47674138747.png)

Allowing the User to easily Create a Tab to put what the user feels are "Advanced" settings for their needs, and that they don't commonly use would allow to clear up the clutter, while not hiding away useful stuff into oblivion.

**Comment by reduz** [comment]
@NathanLovato As it was discussed in #4724 this can be done easily by pinning properties. They are easy to discover and edit because they appear on top, they work like any other property (does not require a custom editor) and you can still copy, paste, drag, animate them, etc. like you do with inspector properties. 

Looks like a far better solution to me that still solves the problem.

**Comment by Jojox** [comment]
> @mrjustaguy To be honest, there is not a lot of clutter, and the proposal keeps the clutter to the same amount as it is now, which is far from problematic.

I agree that this proposal make those tools more explicit anyway so it's a good first step. I missed the change in top bar for a long time when beginning with godot.

The dedicated panel i mentionned could be done later on, maybe it would always be visible and we could transfer:
![image](https://user-images.githubusercontent.com/6182189/159956641-fa1079e2-aad1-4e49-82d4-1ed30f475a71.png)
This would make this panel a kind of "Toolbar" following the same hierarchy than the "Inspector", each set of action would be linked to the selected node types (and handle inheritance without loosing actions)

**Comment by SirLich** [comment]
I think that this needs a proper UX pass, honestly. 

It's likely that the "right" solution is simply making the action-bar (in it's current location) more visible.

**Comment by MewPurPur** [comment]
Hard disagree. I think the drawers PR would be the better path forward for Godot.

I'll go over your framing of the issue and why I think it falls flat.

**1st issue - They mix editing and action**

"Action" is just a type of high-level editing. Control Layouts we call editing because they configure multiple properties at once, but they aren't in the inspector. Add Line2D Point is an action because it requires a click on the viewport, but it still simply configures multiple properties by using information about that click.

The true split should be between low-level properties (Inspector) and high-level editing (contextuals).

A few contextual settings right now are indeed just easy access for low-level properties, like Label's font edit. I agree it would be better to replace _those_ with property pinning. But again, that's just font edit, yet you're talking down drawers which only implement something that already exists.

**2nd issue - something about 3D**

I don't use 3D but I can't imagine what could go so wrong with drawers there. Is this a problem with the idea or something that isn't figured out and implemented yet?

**3rd issue - Contextually disconnected**

If low-level editing is shifted back to the inspector, this fails to hold. This PR should be competing with drawers, not how the already-existing things  that are being implemented inside them.

In fact, if I can drag my drawers around, such as moving a Line2D drawer in the viewport where I want to edit my line, I'd argue it would be _more_ contextually connected.

**4th issue - Folded drawers not clear**

There's only one thing to know. If a node has contextual settings, they will be in small boxes next to the inspector. While action areas are slightly more descriptive at first, they fall off after that. Your example of Line2D editing, for example: I need to hover to see what each of the 3 buttons does, whereas drawers have the space to provide this context.

**5th issue - Floating things is bad**

Ultimately it's about finding a good balance of descriptiveness, convenience, and compactness. Adjusting either of these affects the other three.

If I have to rate this proposal compared to drawers, I must say it's less descriptive (as Iexplained in point 4), it's less convenient (as I explained in point 3) and it's less compact. It touts itself for being more compact, but this is all due to the sister proposal to this one, which I support.

Floating stuff is not ideal. But this proposals has its quirk too - cluttering the inspector with non-properties. Sure, we can remove clutter in other ways, but like folks have said, "The best way to treat a disease is to prevent it". Contextuals should be prevented from cluttering the inspector.

**Comment by reduz** [comment]
@MewPurPur As I mentioned multiple times above, this proposal and contextual drawers do not have much overlap. I feel you are missing the point comparing both. 

What replaces the contextual drawers for the most part is property pinning (https://github.com/godotengine/godot-proposals/issues/4274), which is vastly superior because it uses the existing systems and prevents having to write customized, inflexible and hard to maintain code which are the drawers (explained why its better [here ](https://github.com/godotengine/godot-proposals/issues/4273#issuecomment-1077760347)).

**Comment by SilencedPerson** [comment]
@reduz 
> @Jojox If Godot used a Blender or Unity like UI, your proposal would make sense, but it does not. The general vision in Godot is that things come into view when you need them. The vision in Blender or Unity is that you decide what goes into view.

Somewhat unrelated to the proposal but was there ever any attempt to seek opinion of the userbase on this subject? For example I absolutely hate that this is the case. Godot feels super fiddly, especially when you need to check something inside a script when you are editing a shader and similar contexts like defining sprite animations or basically doing anything in the bottom drawer under the viewport.

**Comment by reduz** [comment]
@SilencedPerson The general consensus is that the way it is now works for the most part, and the UI has seen increased improvements to make workflows easier. That is not to say more editor customization will not happen, but things are being changed little by litte.

**Comment by MewPurPur** [comment]
@reduz I did see both proposals, and as I mentioned, property pinning only covers contextuals that are essentially easy access to low-level properties (i.e. font edit). I, too, want to see those contextuals gone in favor of property pinning.

This PR does, however, compete with drawers on what to do with the remaining contextuals which handle high-level editing. It does it worse.

The point about drawers being less maintainable might be valid, but I fail to see why. Contextuals just get a new UX and can be organized as a basic dropdown by default. Being able to put in extra effort to organize them neatly (like pycbouh organized Control Layouts) just means there's further room for improvement, for a concept that's already better.

**Comment by reduz** [comment]
@MewPurPur It does not make sense to use a drawer for a single button or menu (what most of the contextuals are).

**Comment by neikeq** [comment]
I disagree about floating things being bad. They help to save space and can increase discoverability. Specially with controls that become visible when selecting a node.
They could also be made movable, like in recent versions of Unity which has movable floating toolbars.
Yes, I like to mention Unity, Unreal and Blender because we're not the pinnacle of UX and there's software out there that has great solutions to solve similar problems.

That being said, in case drawers are rejected, I still prefer the current toolbar over this proposal. It's much closer to where it's needed. Think of polygon editing for example. It would be very annoying moving the mouse all the way to the inspector.
We could improve the current toolbar discoverability with some animated effect to highlight it for an instant when it appears.

**Comment by reduz** [comment]
@neikeq We already tried improving the current toolbar discoverability by having a highlight, but its still not great. Polygon editing is probably the exception that may make some sense there (or not since its still pretty un-discoverable) but for all the other remaining ones it makes zero sense for them to be on the toolbar. 

Since you mention Unity, most of these things are in the inspector area in Unity too, so that argument could be made either way.

I don't know what the pinnacle of usability is, all I know is that we must try to do what we find to work by ourselves, not try to copy Unity just for the sake of it.

**Comment by MewPurPur** [comment]
@reduz Formatting Control anchors, for example, is just one click, but it changes multiple values, so I cannot pin properties that completely replace this functionality. This one click is still high-level editing.
The property pinning proposal **does not** cover this gap. Of the contextuals it could fully replace, I can really only think of font edit.

**Comment by reduz** [comment]
@MewPurPur You are right on this, and I fully agree, but we can modify the inspector (or worst case action area) to make this happen and ensure its easy and accessible. The main issue here is that this is more the exception than the rule. 

The font settings in the drawers will cease being needed here because we are discussing a refactor of how this is exposed in a way that makes more sense, so only having the drawers feature _just_ for control alignment does not make a lot of sense. Specially when it can be moved to action area or inspector (and have pinning to have it handy while editing). The rest of the actionables are either a menu or a button, so they don't justify a drawer. 

My feeling on this is that the drawers look nice but are a solution in search of a problem. They are not needed for the most part.

**Comment by neikeq** [comment]
> We already tried improving the current toolbar discoverability by having a highlight, but its still not great.

Is it only the background color? I mean something animated like this:

https://user-images.githubusercontent.com/7718100/160188577-b074b438-081b-4e90-9006-7fb46adc3d42.mp4

> Since you mention Unity, most of these things are in the inspector area in Unity too, so that argument could be made either way.

They used to have a floating tools panel (I think for user tools though, not built-in). But they seem to have replaced it entirely with the floating overlays.

<details>
  <summary>See screenshots</summary>
  
  ![ss1](https://docs.unity3d.com/uploads/Main/ToolmodesPopup.png)
  ![ss2](https://unity3d.com/profiles/unity3d/themes/unity/images/unity/beta/alpha-beta/overlay-configs.gif)
</details>

> I don't know what the pinnacle of usability is, all I know is that we must try to do what we find to work by ourselves, not try to copy Unity just for the sake of it.

What I mean is that when using other software as an example, the mentality seems to be "we're not X, we do things the Godot way". I think we shouldn't be afraid to borrow ideas that work in other projects.

**Comment by reduz** [comment]
@neikeq 

> Is it only the background color? I mean something animated like this:

I understand what you mean, but IMO if something does not work, doubling down on that path is generally a mistake. I am not necessarily against floating panels, but I think in this context it makes no sense to use it for actionable areas.

> What I mean is that when using other software as an example, the mentality seems to be "we're not X, we do things the Godot way". I think we shouldn't be afraid to borrow ideas that work in other projects.

I fully agree with you, but I feel it is you who gets this backwards. If you try to understand better the problem at hand, you will realize that Unity actually solves this precise problem in a similar (but not exactly the same) way to this proposal. As an example, anchoring menu is located in the inspector:

![image](https://user-images.githubusercontent.com/6265307/160190362-926ea5ed-fc6c-4ac6-8336-59a430296781.png)

And so is lightmap baking:

![image](https://user-images.githubusercontent.com/6265307/160190456-4c4442cc-68df-4c0c-acdd-0c05cdbf4d75.png)

And this makes sense usability wise because everything related is put together. Users will never miss something that appeared somewhere else.

In Godot we have these in the Viewport toolbar as actionable content, so there is a disconnection between the object being edited and this actionable content. 

The most natural solution to me is, since have this strong distinction of actionable content and edition is to keep them separate but at least grouped in the same place (like it happens in Unity) and this is what the proposal is about. 

So in the end what's the point on insisting in trying to keep it separate in a toolbar or a floating dock if having it together works better in practice no matter how you look at it? IMO arguments related to "not mixing actions and properties" being good is subjective bikeshedding without much substance.

**Comment by neikeq** [comment]
I think some actions make sense in the inspector. Others don't, like the polygon tools. I think it would be bad to move the latter to the inspector.

> I understand what you mean, but IMO if something does not work, doubling down on that path is generally a mistake. I am not necessarily against floating panels, but I think in this context it makes no sense to use it for actionable areas.

That highlighting animation is meant for the actions in the current toolbar we have.

**Comment by reduz** [comment]
@neikeq

> I think some actions make sense in the inspector. Others don't, like the polygon tools. I think it would be bad to move the latter to the inspector.

I think in general the whole of Polygon is a mess. It could be left as-is for the time being, but it will need some serious rework at some point.

**Comment by neikeq** [comment]
Regarding inspector space, what about hiding the search text field and merging the toolbar? Clicking the search button would display the text field either below or on top of the toolbar:

![image](https://user-images.githubusercontent.com/7718100/160211334-ae73c3f0-331a-4466-9db3-9d89b8e98da7.png)

**Comment by Diddykonga** [comment]
I really like @neikeq suggestion to remove search bar, and only make it show up when clicking the search button.

**One thing that I would really want from this proposal though:**

Allow Scripts/Classes to export functions to the Inspector's action area, this makes the whole action area more general and allows easy extending to do what the developer desires of the object.

**Comment by reduz** [comment]
@Diddykonga believe it or not this is implemented (but not documented). The main problem with this is that its easy to break the undo/redo history if you just call functions on scripts, so its generally better to do this via editor plugins.

**Comment by Nodragem** [comment]
I think it is a good idea to have an action area; and/or to have some oneshot action buttons in the inspector where it makes sense. The example with Unity convinced me.

I like the insight that we have "actions" vs "data" to be displayed in the UI. As @MewPurPur commented, there are low-level "actions" (e.g. moving a slider) that modifies only one variable of the "data"; and there are high-level "actions" that modifies a set of variables (e.g. align to top-left). 

However, I think to complete the story, there is a need of another concept that is that of a "tool" (or interactive action). This is where the special case of the "Polygon Editing" buttons falls in:
- an "action" is a oneshot button that triggers a set of modifications to the data (e.g. change the font color, align to left, etc)
- a "tool" is a temporary interaction mode allowing to "continuously act" on the data in an intuituve way in the viewport (e.g. change text position with the mouse, extrude the face of a polygon, etc)

Hence, my point is that this proposal and its sister proposal are good opportunities to think about the place of interactive actions ("tools") in Godot's UI. Should we have a dedicated "tool area"? I think to have this space would encourage people to create interactive tools that eases complex editing (e.g. creating bones, polygons, path, UI, etc).

**Comment by pycbouh** [comment]
I am against this idea, but I'm staying out of this discussion because I've already given my feedback a couple of months ago and this proposal doesn't present any new information, so it's either going to be accounted for or not 🙃 Besides, you've already been doing a good job highlighting the issues with this idea, such as increased clutter in the inspector.

However, one thing just occurred to me. If this is the way to go, we don't actually need a proposal to try it. Godot already has everything required to add controls at the top of the property list in the inspector. `EditorInspectorPlugin` has two methods, `parse_begin` and `parse_end` that allow to add arbitrary controls at the start and at the end of the entire tree of properties. It's all you need to test drive this idea.

Here's a screenshot of some tests I've been doing last year that demonstrate that this is indeed trivial to do:

![image](https://user-images.githubusercontent.com/11782833/161258949-a5efb30d-1542-4327-b9c8-1d6fe62e0396.png)

So, if this is indeed the way to do it and it's a better place for some tools than drawers or classic toolbars, as this proposal leads us to believe, it should be easy to make a PR to give us hands-on testing for this. That way Godot users can give actual valuable feedback and compare this in practice to having [tool drawers](https://github.com/godotengine/godot/pull/55630) and classic toolbars instead. UX, after all, should be tested by experiencing it.

**Comment by reduz** [comment]
@pycbouh My idea of action area is that it would be placed a bit above so it does not scroll, to make it a bit separate from the data area. I really am not sure if this is better or worse than placing it on the list of properties.

I fully agree with you in the end that we should test this by experiencing it. This proposal is far simpler to implement than the contextual drawers as its for the most part moving toolbar stuff to the top section of the inspector (there are a couple dozen of those and it should be mostly replacing the location and doing some minimal changes), and does not have the drawbacks of not working for 3D as-is.

For the font stuff, I want to add that we have been discussing with Pavel and we are going to do some changes to how this is presented, as it will most likely be going back to the inspector as an optional font settings resource you can set in nodes that make sense (in a way where its clearly visible to the user and grouped with similar properties).

So IMO action area + separate font settings resource should be a short term alternative that is simpler to implement. If we see that this remains confusing to user, I believe it will be a good opportunity to try something more involved like the contextual drawers, or at least having contextual drawers for things that don't work well in the action area.

**Comment by lufog** [comment]
I really would like to see this implemented. Because now editor, in many cases almost unusable, even with decreased display scale:

![image](https://user-images.githubusercontent.com/10379487/162526147-f5cbb986-017c-4a94-96c5-2c1a628a0ae1.png)

**Comment by fire-forge** [comment]
I like @YuriSizov's concept because Godot users are already used to looking at the top of the inspector for tools for editing nodes and resources. Many of them have editors or previews there (Gradient, GradientTexture2D, Mesh, Curve, Material, etc.), so moving the toolbar buttons there would mean there's less places to look for tools.

> My idea of action area is that it would be placed a bit above so it does not scroll, to make it a bit separate from the data area.

Perhaps the action area could be above the scrolling area (like in your concept) for the "main" edited object and at the top of the properties list for sub-inspectors (like in YuriSizov's concept). This would also remove the need for https://github.com/godotengine/godot/pull/52984 and https://github.com/godotengine/godot/pull/45085, since every inspector can have it's own action area.

**Comment by realkotob** [comment]
I suggest we go with this other solution temporarily just so we can fix the cut-off issue.

https://github.com/godotengine/godot-proposals/issues/386

We can then re-structure the UI at our leisure, but for now I think this proposal here will take some time.

**Comment by MewPurPur** [comment]
@lufog Until the issue is resolved, the temporary solution is to make the inspector floating.

**Comment by cassandrus** [comment]
Seems like a lot of controversy here. :/
I'd like to share my opinion as more or less fresh dev/user of Godot.

1. It's hard to get that "actions toolbar" has any relation to node functionality. Overall scarcity of actions there makes it even worse because most of the time there are no special actions there. But suddenly needed Action spawns there on some "special" nodes and it is every time confusing. So, idea to add Actions in a place more related to node seems right in overall. 
2. I reread proposals and it says that only actions go there but every time I see these "Alignment" and "Anchors" as examples. Are these actions at all? Aren't they just doubling Inspector functionality? Really confusing.
3. In my view "viewport toolbar" can store some node functionality but only if it somehow relates to this Viewport (like camera "preview").
4. What makes it funnier is that there is another "actions menu" that is ignored in this conversation. When I click AnimationPlayer node, it opens Animation tool in the footer. When I open ResourcePreloader node, it opens ResourcePreloader tool footer. Are these not actions? As with "actions toolbar" this "click and something will open in footer" functionality is mostly unused, so I just can't get why you can't put all Actions from "actions toolbar" to footer, making footer also spawn for nodes with actions? There is a lot of space and it will work well as user knows that tools are always in the footer.

**Comment by Calinou** [comment]
> I reread proposals and it says that only actions go there but every time I see these "Alignment" and "Anchors" as examples. Are these actions at all? Aren't they just doubling Inspector functionality? Really confusing.

Setting alignment and anchors will configure several properties at once, which is why they're not properties in the inspector. There are a handful [other properties](https://github.com/godotengine/godot/issues/57366) where automatic adjustment of this kind happens, and I don't consider it particularly user-friendly as you may never notice the other properties changing behind your back.

**Comment by QbieShay** [comment]
Adding to the conversation since i just made a mockup, I think it would be useful if these tools would be put (where it makes sense) in the relevant sections of the inspector. Example with particle skeletal emission:
![image](https://github.com/godotengine/godot-proposals/assets/7917475/390de7cd-4651-4f44-ad98-4ecb4fd46b0c)

Blender does the same with baking maps.

**Comment by Calinou** [comment]
@QbieShay This looks pretty similar to https://github.com/godotengine/godot-proposals/issues/2149, so I'd suggest using the same UI for both exported actions and built-in actions.

**Comment by Nodragem** [comment]
I closed my proposal in favour to this one, but I would like to add that we need to distinguish:
- Actions
- Editing modes

It makes sense to Actions like “Clear NavigationMesh” and “Bake NavigationMesh” to make to the Action Area, but Editing modes like “Edit Polygon” should stay at the top of the 3D Viewport as they are Editing modes of the 3D Viewport.

Also, I think we could still have all actions accessible in a menu “nodename actions” in the 3D Viewport top bar. It is not a bad things to be able to access the actions in different ways, and it will cover cases where the inspector is not opened.
